packages/settings-library/src/settings_library/base.py

```diff
@@ -1,74 +1,111 @@
-import logging
-import os
 from functools import cached_property
-from typing import List, Tuple, Type
+from typing import Sequence, get_args
 
-from pydantic import BaseSettings, Extra, SecretStr, ValidationError
+from pydantic import BaseConfig, BaseSettings, Extra, ValidationError, validator
+from pydantic.error_wrappers import ErrorList, ErrorWrapper
+from pydantic.fields import ModelField, Undefined
+from pydantic.types import SecretStr
 
-logger = logging.getLogger(__name__)
+
+class DefaultFromEnvFactoryError(ValidationError):
+    ...
+
+
+def create_settings_from_env(field):
+    # Keeps a reference of field but MUST nothing should be modified there
+    # cannot pass only field.type_ because @prepare_field still not resolved!
+
+    def _default_factory():
+        """Creates default from sub-settings or None (if nullable)"""
+        field_settings_cls = field.type_
+        try:
+            return field_settings_cls()
+
+        except ValidationError as err:
+            if field.allow_none:
+                return None
+
+            def _prepend_field_name(ee: ErrorList):
+                if isinstance(ee, ErrorWrapper):
+                    return ErrorWrapper(ee.exc, (field.name,) + ee.loc_tuple())
+                assert isinstance(ee, Sequence)  # nosec
+                return [_prepend_field_name(e) for e in ee]
+
+            raise DefaultFromEnvFactoryError(
+                errors=_prepend_field_name(err.raw_errors),  # type: ignore
+                model=err.model,
+                # FIXME: model = shall be the parent settings?? but I dont find how retrieve it from the field
+            ) from err
+
+    return _default_factory
 
 
 class BaseCustomSettings(BaseSettings):
-    class Config:
-        # SEE https://pydantic-docs.helpmanual.io/usage/model_config/
-        case_sensitive = False
+    """
+    - Customized configuration for all settings
+    - If a field is a BaseCustomSettings subclass, it allows creating a default from env vars setting the Field
+      option 'auto_default_from_env=True'.
+
+    SEE tests for details.
+    """
+
+    @validator("*", pre=True)
+    @classmethod
+    def parse_none(cls, v, field: ModelField):
+        # WARNING: In nullable fields, envs equal to null or none are parsed as None !!
+        if field.allow_none:
+            if isinstance(v, str) and v.lower() in ("null", "none"):
+                return None
+        return v
+
+    class Config(BaseConfig):
+        case_sensitive = True  # All must be capitalized
         extra = Extra.forbid
         allow_mutation = False
         frozen = True
         validate_all = True
         json_encoders = {SecretStr: lambda v: v.get_secret_value()}
```

Comment:
THOUGHT: I start reading quite a lot about orjson or ujson being many times faster than the usual json library... maybe at some point we should add it too

Reply:
#2760